usr/lib/common/lock_sess_mgr.c

```diff
@@ -865,233 +917,338 @@ CK_RV session_mgr_set_op_state(STDLL_TokData_t *tokdata, SESSION *sess,
+                if (auth_key == CK_INVALID_HANDLE) {
+                    TRACE_ERROR("%s\n", ock_err(ERR_KEY_NEEDED));
+                    return CKR_KEY_NEEDED;
+                }
+                ptr1 = (CK_BYTE *) ctx;
+                ptr2 = ptr1 + sizeof(SIGN_VERIFY_CONTEXT);
+                ptr3 = ptr2 + ctx->context_len;
+
+                if (ctx->context_len) {
+                    context = (CK_BYTE *) malloc(ctx->context_len);
+                    if (!context) {
+                        TRACE_ERROR("%s\n", ock_err(ERR_HOST_MEMORY));
+                        return CKR_HOST_MEMORY;
+                    }
+                    memcpy(context, ptr2, ctx->context_len);
+                }
+
+                if (ctx->mech.ulParameterLen) {
+                    mech_param = (CK_BYTE *) malloc(ctx->mech.ulParameterLen);
+                    if (!mech_param) {
+                        if (context)
+                            free(context);
+                        TRACE_ERROR("%s\n", ock_err(ERR_HOST_MEMORY));
+                        return CKR_HOST_MEMORY;
+                    }
+                    memcpy(mech_param, ptr3, ctx->mech.ulParameterLen);
+                }
+            }
+            break;
+
+        case STATE_DIGEST:
+            {
+                DIGEST_CONTEXT *ctx =
+                    (DIGEST_CONTEXT *)(cur_data + sizeof(OP_STATE_DATA));
+
+                len = sizeof(DIGEST_CONTEXT) + ctx->context_len +
+                                                ctx->mech.ulParameterLen;
+                if (len != op_data->data_len) {
+                    TRACE_ERROR("%s\n", ock_err(ERR_SAVED_STATE_INVALID));
+                    return CKR_SAVED_STATE_INVALID;
+                }
+                ptr1 = (CK_BYTE *) ctx;
+                ptr2 = ptr1 + sizeof(DIGEST_CONTEXT);
+                ptr3 = ptr2 + ctx->context_len;
+
+                if (ctx->context_len) {
+                    context = (CK_BYTE *) malloc(ctx->context_len);
+                    if (!context) {
+                        TRACE_ERROR("%s\n", ock_err(ERR_HOST_MEMORY));
+                        return CKR_HOST_MEMORY;
+                    }
+                    memcpy(context, ptr2, ctx->context_len);
+                }
+
+                if (ctx->mech.ulParameterLen) {
+                    mech_param = (CK_BYTE *) malloc(ctx->mech.ulParameterLen);
+                    if (!mech_param) {
+                        if (context)
+                            free(context);
+                        TRACE_ERROR("%s\n", ock_err(ERR_HOST_MEMORY));
+                        return CKR_HOST_MEMORY;
+                    }
+                    memcpy(mech_param, ptr3, ctx->mech.ulParameterLen);
+                }
+            }
+            break;
+        default:
+            TRACE_ERROR("%s\n", ock_err(ERR_SAVED_STATE_INVALID));
+            return CKR_SAVED_STATE_INVALID;
+        }
+
+        /* clear current state (if active) and copy the new state information */
```

Comment:
This comment is slightly wrong. The current state has already been cleared above. So I'll remove the first first part of the comment.